In [1]:
# helper libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sn

In [2]:
# loading the dataset into a variable(data)
# i have the dataset available in the same path, so i metioned it directly
# else we can provide the path to it
data = pd.read_csv('udemy_courses.csv')
# dataset.shape is used to check the shape of the dataset
# it gives (row, cols )
data.shape

(3678, 12)

Note: We got 3678 rows with 12 cols

### Analysis on data

In [3]:
# after loading the data, we must be curious how the data looks like
# dataset.head() allows you to do it
# by default it returns 5 rows as output
# when we mention any numeric(integer) value as an argument to it
# then it return that many number of rows
# Try it with different values!!!
data.head(10)

course_id                                       course_title  \
0    1070968                 Ultimate Investment Banking Course   
1    1113822  Complete GST Course & Certification - Grow You...   
2    1006314  Financial Modeling for Business Analysts and C...   
3    1210588  Beginner to Pro - Financial Analysis in Excel ...   
4    1011058       How To Maximize Your Profits Trading Options   
5     192870  Trading Penny Stocks: A Guide for All Levels I...   
6     739964  Investing And Trading For Beginners: Mastering...   
7     403100  Trading Stock Chart Patterns For Immediate, Ex...   
8     476268  Options Trading 3 : Advanced Stock Profit and ...   
9    1167710  The Only Investment Strategy You Need For Your...   

                                                 url  is_paid  price  \
0  https://www.udemy.com/ultimate-investment-bank...     True    200   
1      https://www.udemy.com/goods-and-services-tax/     True     75   
2  https://www.udemy.com/financial-modeling-for-b...     True     45   
3  https://www.udemy.com/complete-excel-finance-c...     True     95   
4  https://www.udemy.com/how-to-maximize-your-pro...     True    200   
5  https://www.udemy.com/trading-penny-stocks-a-g...     True    150   
6  https://www.udemy.com/investing-and-trading-fo...     True     65   
7  https://www.udemy.com/trading-chart-patterns-f...     True     95   
8  https://www.udemy.com/day-trading-stock-option...     True    195   
9  https://www.udemy.com/the-only-investment-stra...     True    200   

   num_subscribers  num_reviews  num_lectures               level  \
0             2147           23            51          All Levels   
1             2792          923           274          All Levels   
2             2174           74            51  Intermediate Level   
3             2451           11            36          All Levels   
4             1276           45            26  Intermediate Level   
5             9221          138            25          All Levels   
6             1540          178            26      Beginner Level   
7             2917          148            23          All Levels   
8             5172           34            38        Expert Level   
9              827           14            15          All Levels   

   content_duration   published_timestamp           subject  
0               1.5  2017-01-18T20:58:58Z  Business Finance  
1              39.0  2017-03-09T16:34:20Z  Business Finance  
2               2.5  2016-12-19T19:26:30Z  Business Finance  
3               3.0  2017-05-30T20:07:24Z  Business Finance  
4               2.0  2016-12-13T14:57:18Z  Business Finance  
5               3.0  2014-05-02T15:13:30Z  Business Finance  
6               1.0  2016-02-21T18:23:12Z  Business Finance  
7               2.5  2015-01-30T22:13:03Z  Business Finance  
8               2.5  2015-05-28T00:14:03Z  Business Finance  
9               1.0  2017-04-18T18:13:32Z  Business Finance

In [4]:
# Check the missing values
# this would be obivois step, we need to every row, col values are present
# dataset.isnull(), returns as True/ False values in place null/ not null respectively
# dataset.isnull().sum() will gives the count instead of manually counting all Trues
data.isnull().sum()


course_id              0
course_title           0
url                    0
is_paid                0
price                  0
num_subscribers        0
num_reviews            0
num_lectures           0
level                  0
content_duration       0
published_timestamp    0
subject                0
dtype: int64

Note: here we can observe, we have no null values

In [5]:
# so now we check for the data types
# dataset.dtypes gives the datatype of each col available in the dataset
data.dtypes

course_id                int64
course_title            object
url                     object
is_paid                   bool
price                    int64
num_subscribers          int64
num_reviews              int64
num_lectures             int64
level                   object
content_duration       float64
published_timestamp     object
subject                 object
dtype: object

Note: we have few Catgorical and few Numerical values

In [6]:
# we can use dataset.columns to view the column names
data.columns

Index(['course_id', 'course_title', 'url', 'is_paid', 'price',
       'num_subscribers', 'num_reviews', 'num_lectures', 'level',
       'content_duration', 'published_timestamp', 'subject'],
      dtype='object')

In [7]:
# we can describe the dataset using the dataset.describe()
data.describe()

course_id        price  num_subscribers   num_reviews  num_lectures  \
count  3.678000e+03  3678.000000      3678.000000   3678.000000   3678.000000   
mean   6.759720e+05    66.049483      3197.150625    156.259108     40.108755   
std    3.432732e+05    61.005755      9504.117010    935.452044     50.383346   
min    8.324000e+03     0.000000         0.000000      0.000000      0.000000   
25%    4.076925e+05    20.000000       111.000000      4.000000     15.000000   
50%    6.879170e+05    45.000000       911.500000     18.000000     25.000000   
75%    9.613555e+05    95.000000      2546.000000     67.000000     45.750000   
max    1.282064e+06   200.000000    268923.000000  27445.000000    779.000000   

       content_duration  
count       3678.000000  
mean           4.094517  
std            6.053840  
min            0.000000  
25%            1.000000  
50%            2.000000  
75%            4.500000  
max           78.500000

Note: It caluculate the mean, std and count for the numerical varibales

In [8]:
data.describe(include = 'all')

course_id                   course_title  \
count   3.678000e+03                           3678   
unique           NaN                           3663   
top              NaN  Acoustic Blues Guitar Lessons   
freq             NaN                              3   
mean    6.759720e+05                            NaN   
std     3.432732e+05                            NaN   
min     8.324000e+03                            NaN   
25%     4.076925e+05                            NaN   
50%     6.879170e+05                            NaN   
75%     9.613555e+05                            NaN   
max     1.282064e+06                            NaN   

                                                      url is_paid  \
count                                                3678    3678   
unique                                               3672       2   
top     https://www.udemy.com/build-beautiful-html5-we...    True   
freq                                                    2    3368   
mean                                                  NaN     NaN   
std                                                   NaN     NaN   
min                                                   NaN     NaN   
25%                                                   NaN     NaN   
50%                                                   NaN     NaN   
75%                                                   NaN     NaN   
max                                                   NaN     NaN   

              price  num_subscribers   num_reviews  num_lectures       level  \
count   3678.000000      3678.000000   3678.000000   3678.000000        3678   
unique          NaN              NaN           NaN           NaN           4   
top             NaN              NaN           NaN           NaN  All Levels   
freq            NaN              NaN           NaN           NaN        1929   
mean      66.049483      3197.150625    156.259108     40.108755         NaN   
std       61.005755      9504.117010    935.452044     50.383346         NaN   
min        0.000000         0.000000      0.000000      0.000000         NaN   
25%       20.000000       111.000000      4.000000     15.000000         NaN   
50%       45.000000       911.500000     18.000000     25.000000         NaN   
75%       95.000000      2546.000000     67.000000     45.750000         NaN   
max      200.000000    268923.000000  27445.000000    779.000000         NaN   

        content_duration   published_timestamp          subject  
count        3678.000000                  3678             3678  
unique               NaN                  3672                4  
top                  NaN  2013-01-03T00:55:31Z  Web Development  
freq                 NaN                     2             1200  
mean            4.094517                   NaN              NaN  
std             6.053840                   NaN              NaN  
min             0.000000                   NaN              NaN  
25%             1.000000                   NaN              NaN  
50%             2.000000                   NaN              NaN  
75%             4.500000                   NaN              NaN  
max            78.500000                   NaN              NaN

Note: this will show the both the numerical and categorical values

In [9]:
data['subject'].value_counts()

Web Development        1200
Business Finance       1195
Musical Instruments     680
Graphic Design          603
Name: subject, dtype: int64

# Insights from Data
1. Visualize: 
This allows a unique take on the dataset and can be done in several different ways such as charts, tables, maps, and graphs.
2. Analyse and Interpret what you see:
In this step, ask yourself questions like: What can I see in this image? Is it what I expected? Are there any interesting patterns? What does this mean in the context of the data?
3. Documenting your Insights and steps:
This step of documentation can begin before you have a look at the data. Usually, we have expectations and assumptions about a dataset before we start working with it, and there is a reason on why that particular data was chosen
4. Transform datasets:
This step allows for exploration of more patterns and findings. With the insights formed from previous steps, more questions regarding the data or findings might arise, and further inspection or analysis might be required. This can be done by transformation such as zooming(aggregating data points into a single group), filtering, and outlier removal.

# Data Cleaning and Preparation for Machine Learning
## 0. Understanding the Data
Before we start cleaning data for a machine learning project, it is vital to understand what the data is, and what we want to achieve. Without that understanding, we have no basis from which to make decisions about what data is relevant as we clean and prepare our data.
## 1. Examining the Data Set
In order to examine the data set we need few helper modules, so we'll load them first

In [11]:
# helper libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sn


#### load the dataset
We use numpy and pandas libraries most of the time for array operations and  manipulate tabular data respectively
`pd.read_csv('file name/path')` is used to load the file.

In [12]:
# loading the dataset into a variable(data)
# i have the dataset available in the same path, so i metioned it directly
# else we can provide the path to it
data = pd.read_csv('udemy_courses.csv')

(3678, 12)

Let’s use the pandas `.head()` method to display rows of the  DataFrame, just to make sure we were able to load the dataset properly

In [13]:
data.head()

course_id                                       course_title  \
0    1070968                 Ultimate Investment Banking Course   
1    1113822  Complete GST Course & Certification - Grow You...   
2    1006314  Financial Modeling for Business Analysts and C...   
3    1210588  Beginner to Pro - Financial Analysis in Excel ...   
4    1011058       How To Maximize Your Profits Trading Options   

                                                 url  is_paid  price  \
0  https://www.udemy.com/ultimate-investment-bank...     True    200   
1      https://www.udemy.com/goods-and-services-tax/     True     75   
2  https://www.udemy.com/financial-modeling-for-b...     True     45   
3  https://www.udemy.com/complete-excel-finance-c...     True     95   
4  https://www.udemy.com/how-to-maximize-your-pro...     True    200   

   num_subscribers  num_reviews  num_lectures               level  \
0             2147           23            51          All Levels   
1             2792          923           274          All Levels   
2             2174           74            51  Intermediate Level   
3             2451           11            36          All Levels   
4             1276           45            26  Intermediate Level   

   content_duration   published_timestamp           subject  
0               1.5  2017-01-18T20:58:58Z  Business Finance  
1              39.0  2017-03-09T16:34:20Z  Business Finance  
2               2.5  2016-12-19T19:26:30Z  Business Finance  
3               3.0  2017-05-30T20:07:24Z  Business Finance  
4               2.0  2016-12-13T14:57:18Z  Business Finance

Let’s also use pandas `.shape` attribute to view the number of samples and features we’re dealing with

In [14]:
data.shape

(3678, 12)